src/listen/ldap_sync/active_directory.c

```diff
@@ -0,0 +1,684 @@
+ * @param[in] depth	to indent debug messages, reflecting group nesting
+ * @return true or false result of the group evaluation
+ */
+static bool ldap_filter_group_eval(ldap_filter_t *group, fr_ldap_connection_t *conn, LDAPMessage *msg, int depth)
+{
+	ldap_filter_t	*node = group->child;
+	bool		filter_state = false;
+
+	DEBUG3("%*sEvaluating LDAP filter group %s", depth, "", group->orig);
+	depth += 2;
+	while (node) {
+		switch (node->filter_type) {
+		case LDAP_FILTER_GROUP:
+			filter_state = ldap_filter_group_eval(node, conn, msg, depth);
+			break;
+		case LDAP_FILTER_NODE:
+			filter_state = ldap_filter_node_eval(node, conn, msg, depth);
+			break;
+		}
+
+		/*
+		 *	Short circuit the group depending on the logical operator
+		 *	and the return state of the last node
+		 */
+		if (((group->logic_op == LDAP_FILTER_LOGIC_OR) && filter_state) ||
+		    ((group->logic_op == LDAP_FILTER_LOGIC_AND) && !filter_state)) {
+			break;
+		}
+
+		node = node->next;
+	}
+	filter_state = (group->logic_op == LDAP_FILTER_LOGIC_NOT ? !filter_state : filter_state);
+
+	depth -= 2;
+	DEBUG3("%*sLDAP filter group %s results in %s", depth, "", group->orig, (filter_state ? "TRUE" : "FALSE"));
+	return filter_state;
+}
+
+#define DEBUG_LDAP_ATTR_VAL if (DEBUG_ENABLED3) { \
+	fr_value_box_t	value_box; \
+	fr_ldap_berval_to_value_str_shallow(&value_box, values[i]); \
+	DEBUG3("%*s  Evaluating attribute \"%s\", value \"%pV\"", depth, "", node->attr, &value_box); \
+}
+
+/** Evaluate a single LDAP filter node
+ *
+ * @param[in] node	to evaluate.
+ * @param[in] conn	LDAP connection the message being filtered was returned on.
+ * @param[in] msg	to filter.
+ * @param[in] depth	to indent debug messages, reflecting group nesting.
+ * @return true or false result of the node evaluation.
+ */
+static bool ldap_filter_node_eval(ldap_filter_t *node, fr_ldap_connection_t *conn, LDAPMessage *msg, int depth)
+{
+	struct berval	**values;
+	int		count, i;
+	bool		filter_state = false;
+
+	switch (node->filter_type) {
+	case LDAP_FILTER_GROUP:
+		return ldap_filter_group_eval(node, conn, msg, depth);
+
+	case LDAP_FILTER_NODE:
+		DEBUG3("%*sEvaluating LDAP filter (%s)", depth, "", node->orig);
+		values = ldap_get_values_len(conn->handle, msg, node->attr);
+		count = ldap_count_values_len(values);
+
+		switch (node->op) {
+		case LDAP_FILTER_OP_PRESENT:
+			filter_state = (count > 0 ? true : false);
+			break;
+
+		case LDAP_FILTER_OP_EQ:
+			for (i = 0; i < count; i++) {
+				DEBUG_LDAP_ATTR_VAL
+				if ((node->value->length == values[i]->bv_len) &&
+				    (strncasecmp(values[i]->bv_val, node->value->vb_strvalue, values[i]->bv_len) == 0)) {
+					filter_state = true;
+					break;
+				}
+			}
+			break;
+
+		/*
+		 *	LDAP filters only use one wildcard character '*' for zero or more
+		 *	character matches.
+		 */
+		case LDAP_FILTER_OP_SUBSTR:
+		{
+			char const	*v, *t, *v_end, *t_end;
+			bool		skip;
+
+			t_end = node->value->vb_strvalue + node->value->length - 1;
```

Comment:
Why is one being subtracted here?  Usually in parsing end is set to the char after the last char in the buffer, and we check for `p < end`